Add reStructuredText (ui.rst) element and documentation.

ZeroCool940711 opened the pull request:
This pull request adds a new ReStructuredText (ui.rst) element to the Public API, along with a basic documenation for it. This new element allows for rendering reStructuredText content by converting it from reStructuredText to html so we can show it in NiceGUI, for this we use `docutils.core.publish_parts`.

falkoschindler reviewed (changes requested):
Thanks for this pull request, @ZeroCool940711! This would, indeed, be a handy UI element.

- [x] You seem to have copy-pasted code for using codehilite, which seems to be ineffective for `ui.rst`. I think we should remove it.
- [x] Same for Mermaid: Are you trying to support Mermaid diagrams within `ui.rst`? If not, we should remove it from the implementation.
- [x] Please try not to re-arrange imports. We're using `isort` to automatically sort and arrange them.
- [ ] You can re-use `remove_indentation` from markdown.py to avoid code duplication.

ZeroCool940711 commented:
> * [ ]  You seem to have copy-pasted code for using codehilite, which seems to be ineffective for `ui.rst`. I think we should remove it.
> * [ ]  Same for Mermaid: Are you trying to support Mermaid diagrams within `ui.rst`? If not, we should remove it from the implementation.

The element was copy-pasted from the markdown element, I was expecting some things to not be needed, since I was not sure what I should leave or remove I left everything as it was since it was working but I will remove the unnecessary parts.

> * [ ]  Please try not to re-arrange imports. We're using `isort` to automatically sort and arrange them.

I also use `isort` and the imports are sorted automatically when saving any file, I didn't do any manual sorting of the imports, so I would assume that was `isort` doing its job.

> * [ ]  You can re-use `remove_indentation` from markdown.py to avoid code duplication.

I can do that, I initially copy-pasted the code as it was thinking I might need to modify some parts of it but since most of the code is the same I will remove that code and import it from `markdown.py` to remove code duplication.

ZeroCool940711 commented:
@falkoschindler I've fixed the issues you mentioned with unused code and I imported `remove_indentation` from `markdown.py`, can you take a look and let me know if there is anything else that need fixing.

falkoschindler reviewed (changes requested):
Thanks for the update, @ZeroCool940711!

I reviewed the code and made a few changes here and there - mainly renamed `ui.rst` to `ui.restructured_text` since we try to avoid abbreviations for a more descriptive API.

- [x] One open issue is syntax highlighting: The documentation says you can simply specify a language, but apparently it doesn't work. The generated HTML already contains classes for each token, but there are no style definitions. It would probably be good to support syntax highlighting, I just don't see how at the moment. Do you have an idea?

ZeroCool940711 commented:
> I reviewed the code and made a few changes here and there - mainly renamed `ui.rst` to `ui.restructured_text` since we try to avoid abbreviations for a more descriptive API.

I think `ui.rst` would be easier to type and use but if you think its better to have a long name that is more descriptive then I'm okay with that.

> * [ ]  One open issue is syntax highlighting: The documentation says you can simply specify a language, but apparently it doesn't work. The generated HTML already contains classes for each token, but there are no style definitions. It would probably be good to support syntax highlighting, I just don't see how at the moment. Do you have an idea?

I tried several things but I couldn't fix that issue, I mean, I did kinda fix it but the issue is that if I fixed it for when we use the element it would break the documentation and if I fixed it for the documentation it would break for when you use the element, I couldn't fix both. If you print the `html` variable from `prepare_content` which is what `docutils.core.publish_parts` give us back, you can see that if we use the `html['whole']` we can get more stuff, we get the complete html with css included, if we use that to render the element when using it it works perfectly and we get syntax highlighting but it breaks the layout of the documentation page, we might be able to fix that with custom css similar to what was done to the markdown element but I'm not good with CSS and I couldn't fix it myself.

falkoschindler reviewed (approved):
Ok, I think I managed to fix the syntax highlighting. Since it uses a similar mechanism like `ui.markdown`, I took the opportunity to remove restructured_text.js and simply let `ui.restructured_text` derive from `ui.markdown`.

falkoschindler commented:
For the record: I experimented with the following code.

```py
ui.restructured_text('''
    =================
    This is a heading
    =================

    .. code-block:: python3

        from nicegui import ui

        ui.label('Hello World!')

        ui.run()

    This is a paragraph that contains `a link`_.

    .. _a link: https://domain.invalid/
''')

ui.markdown('''
    # This is a heading

    ```python
    from nicegui import ui

    ui.label('Hello World!')

    ui.run()
    ```

    This is a paragraph that contains [a link](https://domain.invalid/).
''')
```
<img width="370" alt="Screenshot 2024-02-21 at 07 54 35" src="https://github.com/zauberzeug/nicegui/assets/5767091/cf211b76-e658-4d0c-ad0d-17f655660b0d">
